.github/workflows/integration_test.yml

```diff
@@ -0,0 +1,40 @@
+name: "Integration test"
+# NOTE: We're just running the tests that require earthdata login here; we
+#       don't distinguish yet between unit and integration tests yet.
+
+on:
+  pull_request:
+  push:
+    branches:
+      - "main"
+      - "development"
```

Comment:
Given these are slightly "expensive" tests since they download files (and running them for every commit could ultimately influence the icepyx metrics), what are our options for restricting how often they run for a given PR?

Reply:
These triggers (`branches:`) would be for every merge, not for every commit on PRs. We can customize the `pull_request:` triggers. By default, they run on `open` and `synchronize` (effectively, when the PR is pushed to, not on every commit). The full list is here: https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#pull_request

I think the best way we can limit these runs is by adding `paths:` filters. That would enable e.g. docs changes to not trigger the expensive tests!

Reply:
Should we consider this part of this PR, or create an issue to track the work?

Reply:
> (effectively, when the PR is pushed to, not on every commit).

Right - forgot this subtlety - but a PR can have a lot of pushes too!

> Should we consider this part of this PR, or create an issue to track the work?

I would definitely consider it part of this PR. I don't want to have a setup running that will trigger so many runs.

Reply:
>I would definitely consider it part of this PR. I don't want to have a setup running that will trigger so many runs.

Don't want to push back too hard, because I am fine doing it however, but I do want to raise that Travis CI behaves this way already:

https://github.com/icesat2py/icepyx/blob/development/.travis.yml#L21-L24

I would lean towards breaking this work into two PRs: (1) migrate Travis -> GH; (2) more efficient. When those PRs are merged into `development`, they will produce two clearer commits and entries in the changelog. I'm OK with immediately following this PR up with step 2 if you're OK with dividing it!

Reply:
> Travis CI behaves this way already:

Not quite...

>  - name: behind Earthdata
>  if: branch = main OR commit_message =~ nsidc_tests OR type = cron

This isn't running the behind-earthdata tests on the development branch unless "nsidc_tests" is in the commit message. As a result, it only runs for the cron or before releases (when the PR is to main).

Reply:
Good catch, I was being too hasty in my reading! Would you prefer me to make this PR more reflect the Travis setup? What was the cron interval? It's not shown in the config.

Reply:
Cron was weekly. I'm okay changing the config (especially if GH Actions has options that allow us to take a better approach). As we're conversing, I'm recalling that the cron was an effort to catch anything that got messed up through work going into development without having these tests run for every push to every PR to that branch. So I'm happy to nix the cron and have some filters that run it more selectively instead (for PRs to development; we'll probably want to keep it as is for main).

Reply:
What do you think of this filter? https://github.com/icesat2py/icepyx/pull/580/commits/098bb985bbb6b551159ddd23409a3adc900a8d37

Reply:
It still feels to me like it will run a lot (and a lot more than necessary)... I'd like to revisit with fresh eyes tomorrow, but what are your thoughts on something like:
```
on:
  push: # these are ultimately the runs that inform our README status badges and provide the "current status"
    branches:
      - "main"
      - "development"
  pull_request: # we want to know that we're not going to break anything with the release
    branches:
      - "main"
 pull_request_review:
    branches:
      - "development"
    paths:
      (list from above)
```

If I've got it right, we run: (1) once for each release (the push to main); (1) one (or more, but hopefully only one) time when we open a release PR from development to main (the pull_request trigger); and (1-2) once or twice for each PR to development (possibly once after it's approved if a relevant path is touched, the pull_request_review trigger, and once on the development branch for current status (the push to development)). By my calculations, this keeps us under ~50 downloads/month, probably with the exception of the next couple!

Reply:
I've never used `pull_request_review` event before. I think we will only want to trigger on the "submitted" type for that event. Let's try it out! NOTE: This still won't fix our fork problem :)

Reply:
> I think we will only want to trigger on the "submitted" type for that event. Let's try it out! 

Or "approved"? I think this is how I set up the uml diagram action so it only runs once per PR.

> NOTE: This still won't fix our fork problem :)

Wouldn't that be nice...

Reply:
4183ece look OK?

Reply:
>Or "approved"? I think this is how I set up the uml diagram action so it only runs once per PR.

There is no approved event type. That workflow checks the current state of the PR after triggering the workflow, and if it's not "approved", it sets the workflow to "skipped": https://github.com/icesat2py/icepyx/blob/a52bc378ab5817db38cd2b47b5ab0c5352128061/.github/workflows/uml_action.yml#L10

If there was an approved event type, then we could avoid the triggering in the first place :(

But since we have so many other possible triggers we would need something like `${{ github.event.action != "pull_request_review" || github.event.review.state == "approved" }}`. Thoughts?

Reply:
> There is no approved event type.

Guess that was wishful thinking on my part then.

> Since we have other possible triggers we would need something like ${{ github.event.action != "pull_request_review" || github.event.review.state == "approved" }}. Otherwise, integration tests would be skipped on our release PRs until they were approved, and if we want that behavior we can remove the pull_request trigger. Thoughts?

I'm happy with the latest edits versus this. I don't think we tend to have a lot of official "reviews" until we're getting pretty close to merging, and it would be great to have the tests run on release PRs right away.

Reply:
> I'm happy with the latest edits versus this.  I don't think we tend to have a lot of official "reviews" until we're getting pretty close to merging, and it would be great to have the tests run on release PRs right away.

I might need to rescind this. I'm noticing the integration test is running, which makes me think that GH considers a comment (regardless of approval) to be a "review", thus triggering the action for any comment made on the PR.

Reply:
I'm thinking the least unpalatable option at this point is your suggestion to remove the PR trigger and add something like `${{ github.event.action != "pull_request_review" || github.event.review.state == "approved" }}`.

Reply:
Let's try it on for size!